### PR TITLE
video 5988 - Use pc stats directly 

### PR DESCRIPTION
**Pull request by makarandp0**
Preflight test was using webrtc module's getStat api, which returns stats [only for 1st nominated](https://github.com/twilio/twilio-webrtc.js/blob/master/lib/getstats.js#L108) active-candidate-pair, which can result in bytesSent/Received to be reported as 0.

Also the api was found to be not returning round trip times for safari.
Fixed this by refactoring the code to use PeerConnection.getStats directly, and just retrieving the stats needed for preflight.
 

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [ ] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: lib/preflight/getCombinedConnectionStats.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+//   const stats = await pc.getStats();
+//   stats.forEach(stat => {
+//     if (['outbound-rtp', 'remote-inbound-rtp', 'inbound-rtp', 'remote-inbound-rtp', 'candidate-pair'].includes(stat.type)) {
+//       if (typeof stats[statName] === 'undefined') {
+//         foundInReport.add(stat.type + '-' + stat.kind);
+//       }
+//     }
+//   });
+//   return foundInReport;
+// }
```
Comment (charliesantos) [inline lib/preflight/getCombinedConnectionStats.ts:46]: Should we remove this?

File: lib/preflight/getCombinedConnectionStats.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    return report;
+  }, {});
+}
+
+function extractSelectedActiveCandidatePair(stats: RTCStatsReport) : SelectedIceCandidatePairStats | null {
+  let activeCandidatePairStatsFound: RTCIceCandidatePairStats|null = null;
+  stats.forEach(statPair => {
+    if (statPair.type === 'candidate-pair' && statPair.nominated) {
```
Comment (charliesantos) [inline lib/preflight/getCombinedConnectionStats.ts:118]: I think `nominated` is not supported on some browsers. Here's a cross browser solution I did for voice. Let me know if this works or if this is applicable to video. See whole method [here](https://github.com/twilio/twilio-client.js/blob/master/lib/twilio/rtc/stats.js#L56).

```js
const selectedCandidatePairReport = candidatePairs.find(pair =>
      // Firefox
      pair.selected ||
      // Spec-compliant way
      (transport && pair.id === transport.selectedCandidatePairId)
    );
```

File: lib/preflight/getCombinedConnectionStats.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { RTCIceCandidateStats, SelectedIceCandidatePairStats,  } from '../../tsdef/PreflightTypes';
+import type { RTCIceCandidatePairStats } from './rtctypes';
+
+interface RTCStatsReport {
+  [x: string]: any;
+  forEach(callbackfn: (value: any, key: string, parent: RTCStatsReport) => void, thisArg?: any): void;
```
Comment (charliesantos) [inline lib/preflight/getCombinedConnectionStats.ts:7]: This is basically `any` type. If we're not declaring all properties, how about doing `type RTCStatsReport = any` instead?

**Review by charliesantos — CHANGES_REQUESTED** [review]

File: lib/preflight/getCombinedConnectionStats.ts (hunk 2 above)
Comment (makarandp0) [reply lib/preflight/getCombinedConnectionStats.ts:118]: Thanks @charliesantos - I like your solution better, I will update.

File: lib/preflight/getCombinedConnectionStats.ts (hunk 3 above)
Comment (makarandp0) [reply lib/preflight/getCombinedConnectionStats.ts:7]: somehow with with the declaration as you suggested, I see an error:
Parameter 'stat' implicitly has an 'any' type. when I use forEach later...with the forEach explicitely declared I do not see the same error

**Review by charliesantos — APPROVED** [review]
Nice work @makarandp0 !
I noticed there's a stable test failing. Can't tell if it's related to this PR https://app.circleci.com/pipelines/github/twilio/twilio-video.js/4357/workflows/e1f2db5e-f980-4c85-9bf6-85b25fb596d7/jobs/47325

File: lib/preflight/getCombinedConnectionStats.ts (hunk 3 above)
Comment (charliesantos) [reply lib/preflight/getCombinedConnectionStats.ts:7]: Ok

**Comment by makarandp0** [comment]
> Nice work @makarandp0 !
> I noticed there's a stable test failing. Can't tell if it's related to this PR https://app.circleci.com/pipelines/github/twilio/twilio-video.js/4357/workflows/e1f2db5e-f980-4c85-9bf6-85b25fb596d7/jobs/47325

yeah, I have seen those tests fail before and those failures are unrelated.

**Pull request merged**